### PR TITLE
Vendoring in libnetwork 2da2dc055de5a474c8540871ad88a48213b0994f

### LINKED ISSUE #13498: libnetwork: kernel freeze/lock up
kernels  3.18-4.0.1

docker/docker tests will freeze when running master

```
 TESTFLAGS='-check.f TestEventsLimit' make test-integration-cli
```

### PR DESCRIPTION
**Pull request by mrjana**
Fixes #13498 

Signed-off-by: Jana Radhakrishnan mrjana@docker.com

### CODE REVIEW

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 1]
```diff
@@ -4,17 +4,28 @@ import (
...
 
 const prefix = "/var/run/docker/netns"
 
-var once sync.Once
+var (
+	once             sync.Once
+	garbagePathMap   = make(map[string]bool)
+	gpmLock          sync.Mutex
+	gpmWg            sync.WaitGroup
+	gpmCleanupPeriod = 60
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: add unit here `60 * time.Second`

**Comment by jfrazelle** [comment]
upgrading the servers ;) going to give this bad boy a spin

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 2]
```diff
@@ -26,11 +37,56 @@ type networkNamespace struct {
...
+			os.Remove(path)
+		}
+
+		gpmWg.Done()
+	}
+}
+
+func addToGarbagePaths(path string) {
+	gpmLock.Lock()
+	defer gpmLock.Unlock()
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: remove defer and put it at end of function

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 3]
```diff
@@ -26,11 +37,56 @@ type networkNamespace struct {
...
+
+	garbagePathMap[path] = true
+}
+
+func removeFromGarbagePaths(path string) {
+	gpmLock.Lock()
+	defer gpmLock.Unlock()
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: same here

File: vendor/src/github.com/docker/libnetwork/sandbox/sandbox_linux_test.go [hunk 4]
```diff
@@ -31,6 +32,9 @@ func newKey(t *testing.T) (string, error) {
 		return "", err
 	}
 
+	// Set the rpmCleanupPeriod to be low to make the test run quicker
+	gpmCleanupPeriod = 2
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/sandbox_linux_test.go]: add unit

**Comment by tiborvass** [comment]
@mrjana what happens if docker quits before routine cleanup?

**Comment by mrjana** [comment]
@tiborvass when we come back up we cleanup the stale paths in `cleanupNamespaceFiles`

**Comment by jfrazelle** [comment]
panic in the unit-tests

```
+ go test -test.timeout=30m github.com/docker/docker/pkg/httputils
--- FAIL: TestResumableRequestReader (0.00s)
panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: cannot assign requested address [recovered]
	panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: cannot assign requested address

goroutine 7 [running]:
testing.func·006()
	/usr/local/go/src/testing/testing.go:441 +0x181
net/http/httptest.newLocalListener(0x0, 0x0)
	/usr/local/go/src/net/http/httptest/server.go:68 +0x457
net/http/httptest.NewUnstartedServer(0x7fd6135fa110, 0xc20802ad80, 0x7fd6135fa110)
	/usr/local/go/src/net/http/httptest/server.go:96 +0x1f
net/http/httptest.NewServer(0x7fd6135fa110, 0xc20802ad80, 0xac3578)
	/usr/local/go/src/net/http/httptest/server.go:83 +0x32
github.com/docker/docker/pkg/httputils.TestResumableRequestReader(0xc20800c120)
	/go/src/github.com/docker/docker/pkg/httputils/resumablerequestreader_test.go:18 +0xe5
testing.tRunner(0xc20800c120, 0xa9d9a0)
	/usr/local/go/src/testing/testing.go:447 +0xbf
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xa8b

goroutine 1 [chan receive]:
testing.RunTests(0x7c25a0, 0xa9d9a0, 0x2, 0x2, 0x100000001)
	/usr/local/go/src/testing/testing.go:556 +0xad6
testing.(*M).Run(0xc20802e460, 0xac2e40)
	/usr/local/go/src/testing/testing.go:485 +0x6c
main.main()
	github.com/docker/docker/pkg/httputils/_test/_testmain.go:102 +0x291

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

Tests failed: ./pkg/httputils

+ go test -test.timeout=30m github.com/docker/docker/pkg/ioutils
PASS
coverage: 79.6% of statements

+ go test -test.timeout=30m github.com/docker/docker/
```

**Comment by jfrazelle** [comment]
so running this binary on the servers and testing this pr with it, i get

```
... Panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: cannot assign requested address (PC=0x43D7E5)
```

in like everyother test-integration-cli test

**Comment by LK4D4** [comment]
If GC will occur right after `syscall.Umount` what difference will be with old behavior?

**Comment by mrjana** [comment]
@LK4D4 it's not that remove by itself is a problem. It's that when we make the kernel interleave between create namespace path and remove is when this problem happens

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
:confetti_ball:

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 5]
```diff
@@ -26,11 +37,54 @@ type networkNamespace struct {
...
+		}
+
+		gpmWg.Done()
+	}
+}
+
+func addToGarbagePaths(path string) {
+	gpmLock.Lock()
+	garbagePathMap[path] = true
+	defer gpmLock.Unlock()
```
Comment (LK4D4) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: what? :)

Comment (vieux) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: :+1:

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 6]
```diff
@@ -26,11 +37,54 @@ type networkNamespace struct {
...
+	// cleanup any stale namespace files if any
+	cleanupNamespaceFiles()
+
+	// Start the garbage collection go routine
+	go removeUnusedPaths()
+}
+
+func removeUnusedPaths() {
+	for {
+		time.Sleep(time.Duration(gpmCleanupPeriod))
```
Comment (LK4D4) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: I think it's better to do with `time.Timer`.

Comment (mrjana) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: I didn't get it. Sorry. :-) 

You want me to wait on  channel returned from `time.After`?

Comment (LK4D4) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: I meant `time.Ticker` :)

Comment (vieux) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: Something like should work.

```
for range time.Tick(gpmCleanupPeriod) {
  ...
}
```

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go (hunk 5 above)
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: @mrjana this should be without the defer :)

Comment (mrjana) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: Yeah taking care of it :-)

**Comment by mrjana** [comment]
@tianon @LK4D4 @vieux updated the PR taking care of the comments

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 7]
```diff
@@ -26,11 +37,52 @@ type networkNamespace struct {
...
+}
+
 func createBasePath() {
 	err := os.MkdirAll(prefix, 0644)
 	if err != nil && !os.IsExist(err) {
 		panic("Could not create net namespace path directory")
 	}
+
+	// cleanup any stale namespace files if any
+	cleanupNamespaceFiles()
```
Comment (icecrime) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: I'm not sure it's a supported use case, but I'm pretty sure reading some people start multiple daemons, and in this case this call might be very problematic.

**Comment by icecrime** [comment]
I think the cleanup logic is fine, but I'm afraid somebody, somewhere, will complain that it breaks running multiple daemon on a single host (was that @SvenDowideit?). I wish we didn't need in-memory state to manage collection, but I don't see an easy to query files to determine if they can safely be removed.

LGTM with a less than 100% confidence ;-) It's already great news that you managed to workaround those kernel bugs: that unblocks 1.7.0-rc1, we can refine if the cleanup logic becomes problematic.

**Comment by mrjana** [comment]
@icecrime if the multiple daemons in a single host is a supported configuration then this could be problematic. Let me think about it a bit more. May be we can rework this a bit after 1.70-rc1?

File: vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go [hunk 8]
```diff
@@ -4,17 +4,28 @@ import (
...
 	"syscall"
+	"time"
 
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/docker/pkg/reexec"
 	"github.com/vishvananda/netlink"
 	"github.com/vishvananda/netns"
 )
 
 const prefix = "/var/run/docker/netns"
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: could this be settable? I wonder whether `--exec-root` which defaults to `/var/run/docker` should set this

Comment (mrjana) [inline vendor/src/github.com/docker/libnetwork/sandbox/namespace_linux.go]: Right now this is not settable. We can make this settable as an enhancement in the future

**Comment by icecrime** [comment]
So now the trade-off is that uncleanly exiting the daemon will leave dangling files behind. Not ideal, but good enough for 1.7.0?

Still LGTM.

**Comment by vieux** [comment]
LGTM

**Comment by vieux** [comment]
It'll be improved in RC2

**Comment by mrjana** [comment]
@icecrime but only for those container which haven't been restarted and only until a system reboot. But may be we can make it better

**Comment by icecrime** [comment]
Alright, let's do this ;-)

**Comment by tiborvass** [comment]
```
----------------------------------------------------------------------
FAIL: docker_cli_run_unix_test.go:162: DockerSuite.TestRunDeviceDirectory

docker_cli_run_unix_test.go:168:
    c.Fatal(err, out)
... Error: exit status 1Error response from daemon: Cannot start container 73017df47d63f1c30f8af70f742fef0ba9f7ef3c45a4431d78c1e8a289b0a9e6: error gathering device information while adding custom device "/dev/snd": lstat /dev/snd: no such file or directory
```

not sure if its related

**Comment by mrjana** [comment]
@tiborvass that happens when you don't have /dev/snd in your machine

**Comment by tiborvass** [comment]
```
----------------------------------------------------------------------
FAIL: docker_cli_nat_test.go:61: DockerSuite.TestNetworkNat

docker_cli_nat_test.go:71:
    c.Fatalf("Failed to connect to server: %v (output: %q)", err, string(out))
... Error: Failed to connect to server: exit status 1 (output: "nc: timed out\n")
```

this is another one

**Comment by jfrazelle** [comment]
test network nat is super unstable it fails all the time for me

**Pull request merged**